### PR TITLE
:sparkles: Add foundation tooling and files

### PR DESCRIPTION
**Pull request by vincepri**
This patchset adds a bunch of basic tools and files needed to start
working on the controllers within the cluster api ecosystem. Enables us
to speed through the basic chores to generate files, run prow jobs
(these still need to be setup at the time of writing), add all the
release tooling and automatic image building.

Signed-off-by: Vince Prignano <vincepri@vmware.com>

### CODE REVIEW

**Review by christopherhein — COMMENTED** [review]
Is the `release-notes` binary supposed to be In the repo, is it managed somewhere else and easily updated?

**Comment by vincepri** [comment]
Yes, it's in Cluster API itself and it's updated by updating the hash version of the capi hack/tools module in `hack/tools/go.mod`

**Comment by christopherhein** [comment]
Perfect, thanks @vincepri should it be loaded at test run or local setup time? Happy to keep it just want to confirm since it's a 2+mb binary in a repo.

/hold for @vincepri to diff another provider

**Comment by vincepri** [comment]
Oh, missed the .gitignore 😂

**Comment by christopherhein** [comment]
> Oh, missed the .gitignore 😂

That makes sense 👍

**Comment by vincepri** [comment]
Should be good to go!

File: PROJECT [hunk 1]
```diff
@@ -0,0 +1,3 @@
+version: "2"
```
Comment (christopherhein) [inline PROJECT:1]: We'll be fast following to change this to `3` once we get https://github.com/kubernetes-sigs/kubebuilder/milestone/21 finished. We might even start this with the `3-alpha`.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	klog.InitFlags(nil)
+
+	_ = clientgoscheme.AddToScheme(scheme)
+	_ = clusterv1.AddToScheme(scheme)
+	// +kubebuilder:scaffold:scheme
+}
+
+// InitFlags initializes the flags.
+func InitFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&metricsAddr, "metrics-addr", ":8080",
```
Comment (christopherhein) [inline main.go:66]: We're going to fast follow to change this to `metrics-bind-address`

File: main.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		"The address the metric endpoint binds to.")
+
+	fs.BoolVar(&enableLeaderElection, "enable-leader-election", false,
```
Comment (christopherhein) [inline main.go:69]: We're going to fast follow to change this to `leader-elect`

File: main.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+		"Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.")
+
+	fs.DurationVar(&leaderElectionLeaseDuration, "leader-election-lease-duration", 15*time.Second,
```
Comment (christopherhein) [inline main.go:72]: Fast follow change this to `leader-elect-lease-duration`

File: main.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+		"Interval at which non-leader candidates will wait to force acquire leadership (duration string)")
+
+	fs.DurationVar(&leaderElectionRenewDeadline, "leader-election-renew-deadline", 10*time.Second,
```
Comment (christopherhein) [inline main.go:75]: fast follow change this to `leader-elect-renew-deadline`

File: main.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+		"Duration that the leading controller manager will retry refreshing leadership before giving up (duration string)")
+
+	fs.DurationVar(&leaderElectionRetryPeriod, "leader-election-retry-period", 2*time.Second,
```
Comment (christopherhein) [inline main.go:78]: Fast follow change this to `leader-elect-retry-period`

**Comment by christopherhein** [comment]
Awesome, thanks for putting all this base work in @vincepri !

/lgtm
/approve

**Comment by vincepri** [comment]
/hold cancel

🎉

**Pull request merged**